### ISSUE TITLE
841 WeaponDefs is zero indexed

### ISSUE THREAD

**Issue by GoogleFrog** [beyond-all-reason/spring#509]
We only discovered this in ZK by cleaning out unused weapons. Now a weapon that appears in the game has weaponDefID 0 and it is causing issues. UnitDefs is one indexed.

```
for i, data in pairs(WeaponDefs) do
	Spring.Echo("index", i, "weapon", data.name)
end

for i, data in pairs(UnitDefs) do
	Spring.Echo("index", i, "unit", data.name)
end
...
[t=00:02:04.208972][f=0001965] index, 703, weapon, zenith_meteor_aim
[t=00:02:04.208998][f=0001965] index, 704, weapon, zenith_meteor_float
[t=00:02:04.209024][f=0001965] index, 705, weapon, zenith_meteor_uncontrolled
[t=00:02:04.209068][f=0001965] index, 0, weapon, amphaa_missile
```

**Comment by badosu** [comment]
Indeed it seems that both [feature](https://github.com/beyond-all-reason/spring/blob/72b2d55d4d4d4747e3b0bf1880516c2fec019af3/rts/Sim/Features/FeatureDefHandler.cpp#L77) and [weapon](https://github.com/beyond-all-reason/spring/blob/72b2d55d4d4d4747e3b0bf1880516c2fec019af3/rts/Sim/Weapons/WeaponDefHandler.cpp#L33) defs are 0-indexed while [unitdefs](https://github.com/beyond-all-reason/spring/blob/72b2d55d4d4d4747e3b0bf1880516c2fec019af3/rts/Sim/Units/UnitDefHandler.cpp#L41) purposefully adds a mock vector entry just so size calculation fits with [1-index](https://github.com/beyond-all-reason/spring/blob/72b2d55d4d4d4747e3b0bf1880516c2fec019af3/rts/Sim/Units/UnitDefHandler.cpp#L64)

**Comment by badosu** [comment]
How can one reproduce the crash in zero-k?

**Comment by badosu** [comment]
- Set my local Zero-K repo to the earlier commit to the amphaa fix
- Gave myself amphaa
- Gave to team 1 planecon
- amphaa shot and killed planecon

I wasn't able to reproduce an issue, although certainly standardizing indexing is desirable it would be good to see a pathological case derived from current behavior.

**Comment by badosu** [comment]
Ok, I was able to identify a pathological case where the developer assumes (in their correct mind) that weapondefs can be iterated exactly like unit and feature defs.

**Comment by badosu** [comment]
To drive home the case, not only ZK but BAR also has a [potential bug](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/359efdf874f24ccb97b7cd7f7d799a9ad8f85586/luarules/gadgets/lups_shockwaves.lua#L22-L24) originating from this discrepancy:

**Comment by GoogleFrog** [comment]
The gadget crash reported to me happened with Angler hit a shield, but there would be many many more silent bugs downstream of this.

**Comment by badosu** [comment]
Fixed on https://github.com/beyond-all-reason/spring/pull/512

**Comment by badosu** [comment]
Reopened for a proper implementation

**Comment by Krogoth100** [comment]
Just adding my 2 cents here
Discovered it a long time ago myself and was surprised. Have been iterating it from 0 since then.
If it gets fixed, it should get a notation in "breaking changes" part of the changelog.

**Comment by badosu** [comment]
True, can you provide the repo you have it so I can create an issue and possible PR?

**Comment by sprunk** [comment]
What was the problem with the first implementation?

**Comment by badosu** [comment]
I missed adjusting the places where AI uses 0-indexed weapons, in principle it should be somewhat straightforward to address.
